drivers/serial/uart_b91.c

```diff
@@ -0,0 +1,568 @@
+static int uart_b91_configure(const struct device *dev,
+			      const struct uart_config *cfg)
+{
+	uint16_t divider;
+	uint8_t bwpc;
+	uint8_t parity;
+	uint8_t stop_bits;
+
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+
+	/* check parity */
+	if (cfg->parity == UART_CFG_PARITY_NONE) {
+		parity = UART_PARITY_NONE;
+	} else if (cfg->parity == UART_CFG_PARITY_ODD) {
+		parity = UART_PARITY_ODD;
+	} else if (cfg->parity == UART_CFG_PARITY_EVEN) {
+		parity = UART_PARITY_EVEN;
+	} else {
+		return -ENOTSUP;
+	}
+
+	/* check stop bits */
+	if (cfg->stop_bits == UART_CFG_STOP_BITS_1) {
+		stop_bits = UART_STOP_BIT_1;
+	} else if (cfg->stop_bits == UART_CFG_STOP_BITS_1_5) {
+		stop_bits = UART_STOP_BIT_1P5;
+	} else if (cfg->stop_bits == UART_CFG_STOP_BITS_2) {
+		stop_bits = UART_STOP_BIT_2;
+	} else {
+		return -ENOTSUP;
+	}
+
+	/* check flow control */
+	if (cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE) {
+		return -ENOTSUP;
+	}
+
+	/* UART configure */
+	uart_b91_cal_div_and_bwpc(cfg->baudrate, sys_clk.pclk * 1000 * 1000, &divider, &bwpc);
+	uart_b91_init(uart, divider, bwpc, parity, stop_bits);
+
+	/* save configuration */
+	DEV_DATA(dev)->cfg = *cfg;
+
+	return 0;
+}
+
+/* API implementation: config_get */
+static int uart_b91_config_get(const struct device *dev,
+			       struct uart_config *cfg)
+{
+	*cfg = DEV_DATA(dev)->cfg;
+
+	return 0;
+}
+
+/* API implementation: driver initialization */
+static int uart_b91_driver_init(const struct device *dev)
+{
+	uint16_t divider = 0u;
+	uint8_t bwpc = 0u;
+	const struct device *pinmux;
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	const struct uart_b91_config *cfg = GET_CFG(dev);
+
+	pinmux = DEVICE_DT_GET(DT_NODELABEL(pinmux));
+	if (!device_is_ready(pinmux)) {
+		return -ENODEV;
+	}
+
+	for (int i = 0; i < cfg->pinctrl_list_size; i++) {
+		pinmux_pin_set(pinmux, B91_PINMUX_GET_PIN(cfg->pinctrl_list[i]),
+			       B91_PINMUX_GET_FUNC(cfg->pinctrl_list[i]));
+	}
+
+	uart_b91_cal_div_and_bwpc(cfg->baud_rate, sys_clk.pclk * 1000 * 1000, &divider, &bwpc);
+	uart_b91_init(uart, divider, bwpc, UART_PARITY_NONE, UART_STOP_BIT_1);
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	cfg->pirq_connect();
+#endif
+
+	return 0;
+}
+
+/* API implementation: poll_out */
+static void uart_b91_poll_out(const struct device *dev, uint8_t c)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	while (uart_b91_get_tx_bufcnt(uart) >= UART_TX_BUF_CNT) {
+	};
+
+	uart->data_buf[data->tx_byte_index] = c;
+	data->tx_byte_index = (data->tx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
+}
+
+/* API implementation: poll_in */
+static int uart_b91_poll_in(const struct device *dev, unsigned char *c)
+{
+	volatile struct uart_b91_t *uart = GET_UART(dev);
+	struct uart_b91_data *data = DEV_DATA(dev);
+
+	*c = uart->data_buf[data->rx_byte_index];
+	data->rx_byte_index = (data->rx_byte_index + 1) % ARRAY_SIZE(uart->data_buf);
+
+	return 0;
+}
```

Comment:
This implementation does not look sane.

`poll_in` is supposed to return data if there is data available in the RX buffer. This implementation is missing any such checks.

https://github.com/zephyrproject-rtos/zephyr/blob/622a8e825658e91e4afed8d6986cc97b81229484/include/drivers/uart.h#L661-L663

Reply:
Fixed.

	if (uart_b91_get_rx_bufcnt(uart) == 0) {
		return -1;
	}